### ISSUE TITLE
Use multi-variant enum as an extrinsic parameter

### ISSUE THREAD

**Issue by anakornk** [polkadot-js/apps#1958]
I'm having a problem with using multi-variant enum as an extrinsic parameter. In this example, my enum is ValueType and i'm trying to call set_enum which has ValueType enum as one of its parameters through polkadot js api. When the enum only has one variant (any of the 4), all works fine. I could also combine uint32 and uint64. However, the other combinations will report an error. For instance, when i tried having all 4 variants together, I could submit an extrinsic with uint32 and uint64, but using either Bool or Char32 will give me an error.

Enum
```
#[derive(Clone, Encode, Decode, PartialEq, Eq, RuntimeDebug)]
pub enum ValueType {
	Char32([u8; 32]),	
	Uint32(u32),
	Uint64(u64),
	Bool(bool),
}
```
```
		fn set_enum(origin, value: ValueType) -> Result {
			// Check if signed
			let sender = ensure_signed(origin)?;
			Ok(())
		}
```

Call set_enum from polkadotjs api
```
api.tx.content
    .setEnum({
      char32: "0x12345678123456781234567812345678"
    })
    .signAndSend(alicePair, { nonce }, ({ events = [], status }) => { ... }
```

Error
```
2019-11-28 12:26:15        RPC-CORE: submitAndWatchExtrinsic(extrinsic: Extrinsic): ExtrinsicStatus:: 1002: Verification Error: Execution(ApiError("Could not convert parameter `tx` between node and runtime: Error decoding field Call :: Content.0")): RuntimeApi("Execution(ApiError(\"
2019-11-28 12:26:15             DRR: Error: submitAndWatchExtrinsic(extrinsic: Extrinsic): ExtrinsicStatus:: 1002: Verification Error: Execution(ApiError("Could not convert parameter `tx` between node and runtime: Error decoding field Call :: Content.0")): RuntimeApi("Execution(ApiError(\"
    at errorHandler (/Users/anakornk/Code/graph4w3/graph-js-sdk/node_modules/@polkadot/rpc-core/index.js:259:26)
    at /Users/anakornk/Code/graph4w3/graph-js-sdk/node_modules/@polkadot/rpc-core/index.js:233:9
    at processTicksAndRejections (internal/process/task_queues.js:82:5)
```

Seems like there is something wrong with the encoding/decoding part. Am i missing anything?

**Comment by jacogr** [comment]
Transferred, since there seems to be an Enum issue with params.

@anakornk Could you please paste the full console logs as to when this issue happens - that will aid in understanding why.

**Comment by anakornk** [comment]
@jacogr 
Here are the steps i did to reproduce the error.
1. ValueType definition in rust.
```
#[derive(Clone, Encode, Decode, PartialEq, Eq, RuntimeDebug)]
pub enum ValueType {
	Char32([u8; 32]),	
	Uint64(u64),
	Bool(bool),
}
```
2. Registered ValueType on Settings->Developer Tab
```
  "ValueType": {
    "_enum": {
      "Char32": "[u8; 32]",
      "Uint64": "u64",
      "Bool": "bool"
    }
  }
```
3.  When i chose the set_enum extrinsinc, a warning is shown. Everything still functions correctly. 
```
Cannot find component for [u8;32], defaulting to Unknown
```
4. When i change Char32 to either Uint64 or Bool, the whole screen goes white. The console logs the following 17 times 
```
backend.js:6 Unable to determine default type for {"info":11,"name":"Char32","type":"[u8;32]","ext":{"length":32,"type":"u8"}}
```
Then ends with a react error
``` 
backend.js:6 Error: Minified React error #185; visit https://reactjs.org/docs/error-decoder.html?invariant=185 for the full message or use the non-minified dev environment for full errors and additional helpful warnings.
```
![image](https://user-images.githubusercontent.com/8156002/69880725-ba533180-1305-11ea-9d84-d14588301851.png)

<img width="1425" alt="Screen Shot 2019-11-30 at 00 06 51" src="https://user-images.githubusercontent.com/8156002/69880689-9394fb00-1305-11ea-8ec3-2253cf25c6e3.png">

**Comment by jacogr** [comment]
There are notoriously hard to track down, however I have access to an unlaunched chain with a bunch of enums, nested and otherwise not.

Just have not had a gap in the last day, but this is high on the list to fix - first thing to do next.

**Comment by anakornk** [comment]
@jacogr I think i've found the problem. It got to do with Dropdown.tsx in the react-components folder. 
_onChange was called twice, once by the onChange handler and the other by the useEffect hook.  
onChange handler first calls _onChange which updates the the state, which then calls the effect hook.

There is no value prop passed down by the EnumParam component, so value is undefined and newValue equals the old/default value. Hence, the second call by the effect hook rewrite the old value to the state causing the state to remain the same (cannot change the enum variant).

packages/react-components/src/Dropdown.tsx 
```
 useEffect((): void => {
    const newValue = isUndefined(value)
      ? defaultValue
      : value;

    // only update parent if we have had something changed
    if (JSON.stringify({ v: newValue }) !== JSON.stringify({ v: stateValue })) {
      _onChange(null, { value: newValue });
    }
  }, [defaultValue, stateValue, value]);
```

I tried adding a condition so that the effect hook was not called when inside EnumParam. It fixed the white screen and can now submit extrinsics with multi-variant enum as a parameter.

**Comment by jacogr** [comment]
Thank you. Feel free to make a PR, otherwise I'll do so in (my) morning and credit you :)

**Comment by anakornk** [comment]
> Thank you. Feel free to make a PR, otherwise I'll do so in (my) morning and credit you :)

 I'm not quite sure what is the best way to fix the problem since i'm not quite familiar with the project yet. I passed down a prop from EnumParam. Maybe there're better ways to do it. Also, got to make sure that the effect will still be called to set the default value. Maybe it'll be better if you fix it, I can help you with testing :)

**Comment by jacogr** [comment]
Will have a fix tomorrow morning EU time, will link here.

Sorry this has taken so long to get to it, was a bit snowed under with Kusama CC3 and supporting the latest Substrate master over the last 3 days.

**Comment by jacogr** [comment]
Ok, based on my tests after #1964, I don't get crashes. However for the multi-variant enum now it doesn't actually change the value, i.e. the enum type. So there is another issue around this Enum component with params. Frustrating, but will take a look at that next.... baby steps.

**Comment by anakornk** [comment]
After the latest fix, everything works fine.
